core/src/main/java/org/hisp/dhis/android/core/trackedentity/TrackedEntityAttributeReservedValueManager.kt

```diff
@@ -0,0 +1,406 @@
+                            getFillUpToValue(
+                                null, trackedEntityAttribute.uid()
+                            )
+                        )
+                    reservedValueSummaries.add(builder.build())
+                }
+            } else {
+                builder.count(blockingCount(trackedEntityAttribute.uid(), null))
+                    .numberOfValuesToFillUp(getFillUpToValue(null, trackedEntityAttribute.uid()))
+                reservedValueSummaries.add(builder.build())
+            }
+        }
+        return reservedValueSummaries
+    }
+
+    private fun increaseProgress(): D2Progress {
+        return d2ProgressManager.increaseProgress(
+            TrackedEntityAttributeReservedValue::class.java, false
+        )
+    }
+
+    private suspend fun downloadValuesForOrgUnits(
+        attribute: String,
+        numberOfValuesToFillUp: Int?
+    ): Flow<D2Progress> = flow {
+        val pattern = trackedEntityAttributeStore.selectByUid(attribute)!!.pattern()
+
+        if (isOrgunitDependent(pattern)) {
+            val organisationUnits = getOrgUnitsLinkedToAttribute(attribute)
+
+            for (organisationUnit in organisationUnits) {
+                downloadValuesIfBelowThreshold(
+                    attribute, organisationUnit, numberOfValuesToFillUp, true
+                )
+                emit(increaseProgress())
+            }
+        } else {
+            downloadValuesIfBelowThreshold(attribute, null, numberOfValuesToFillUp, true)
+
+            emit(increaseProgress())
+        }
+    }
+
+    private suspend fun downloadValuesIfBelowThreshold(
+        attribute: String,
+        organisationUnit: OrganisationUnit?,
+        minNumberOfValuesToHave: Int?,
+        storeError: Boolean
+    ): Unit = coroutineScope {
+        // Using local date. It's not worth it to make a system info call
+        store.deleteExpired(Date())
+        val fillUpTo = getFillUpToValue(minNumberOfValuesToHave, attribute)
+        val pattern = trackedEntityAttributeStore.selectByUid(attribute)!!.pattern()
+        val remainingValues = store.count(
+            attribute, if (isOrgunitDependent(pattern)) getUidOrNull(organisationUnit) else null, pattern
+        )
+
+        // If number of values is explicitly specified, we use that value as threshold.
+        val minNumberToTryFill = minNumberOfValuesToHave ?: (fillUpTo!! * FACTOR_TO_REFILL).toInt()
+
+        if (remainingValues < minNumberToTryFill) {
+            val numberToReserve = fillUpTo!! - remainingValues
+            downloadValues(
+                attribute, organisationUnit, numberToReserve, pattern, storeError
+            )
+        }
+    }
+
+    private suspend fun downloadValues(
+        trackedEntityAttributeUid: String,
+        organisationUnit: OrganisationUnit?,
+        numberToReserve: Int,
+        pattern: String?,
+        storeError: Boolean
+    ) {
+        return suspendCoroutine {
```

Comment:
It seems that this `return` makes the call to never end. I think you can remove this whole line.